 import libraries

In [3]:
import pandas as p
import numpy as np
import pickle
import emoji

  import pickle 
 *  *pickle module implements binary protocols for serializing and de-serializing a* 
    *python object structure. “pickling” is the process whereby a python object hierarchy is*
    *converted into a byte stream*

 let's load some data!   

In [4]:
df=p.read_csv('ZHIV_zip_.csv')
df.head(3)

RegionID  SizeRank  RegionName RegionType StateName State      City  \
0     61639         0       10025        Zip        NY    NY  New York   
1     84654         1       60657        Zip        IL    IL   Chicago   
2     61637         2       10023        Zip        NY    NY  New York   

                         Metro       CountyName  2000-01-31  ...  2021-10-31  \
0  New York-Newark-Jersey City  New York County    329164.0  ...   1153364.0   
1     Chicago-Naperville-Elgin      Cook County    311718.0  ...    523727.0   
2  New York-Newark-Jersey City  New York County    510209.0  ...   1517150.0   

   2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
0   1152736.0   1153314.0   1159292.0   1171216.0   1190200.0   1207107.0   
1    526511.0    528499.0    529879.0    530092.0    532758.0    534840.0   
2   1521442.0   1521759.0   1532449.0   1542269.0   1559390.0   1572653.0   

   2022-05-31  2022-06-30  2022-07-31  
0   1221417.0   1227148.0   1234232.0  
1    539859.0    543658.0    546709.0  
2   1591368.0   1600569.0   1607770.0  

[3 rows x 280 columns]

 export metaData w pickles  🥒 

*  removing region data- iloc[:, 0:9]
*  check column (cols)
*     import os.mkdir()
   *  *mkdir is used for creating or making a new directory*
*     dump metaData
  *  'wb' for writing binary file. 
  *  'fp' file pointer

In [5]:
meta = df.iloc[:, 0:9]

In [6]:
meta.columns

Index(['RegionID', 'SizeRank', 'RegionName', 'RegionType', 'StateName',
       'State', 'City', 'Metro', 'CountyName'],
      dtype='object')

In [7]:
#import os
#os.mkdir('pickle')

In [9]:
with open('pickle/meta.pkl', 'wb') as fp:
    pickle.dump(meta, fp)

 subsetting 
*  creating sac (sacramento) dataFrame
*  get list
*  remove col RegionName
*  drop cols (columns)


In [11]:
sac = df[df.CountyName == 'Sacramento County']

In [12]:
meta_cols = list(df.columns[0:9])

In [13]:
meta_cols.remove('RegionName')

In [14]:
sac = sac.drop(meta_cols, axis = 1)

 examine data 

In [15]:
sac.shape

(50, 272)

In [16]:
sac.head(10)

RegionName  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
102        95630    235832.0    236836.0    237997.0    240134.0    242395.0   
163        95608    174288.0    175027.0    175987.0    177583.0    179274.0   
187        95823    114478.0    114963.0    115594.0    116743.0    117922.0   
376        95758    172491.0    173242.0    174209.0    175923.0    177760.0   
506        95670    151622.0    152162.0    152938.0    154300.0    155740.0   
776        95624    188868.0    189591.0    190462.0    192080.0    193869.0   
1045       95831    183734.0    184517.0    185582.0    187258.0    189063.0   
1066       95828    127990.0    128557.0    129151.0    130358.0    131677.0   
1153       95628    194236.0    195146.0    196134.0    197833.0    199806.0   
1187       95610    151719.0    152299.0    152998.0    154388.0    156028.0   

      2000-06-30  2000-07-31  2000-08-31  2000-09-30  ...  2021-10-31  \
102     244414.0    247327.0    250195.0    253888.0  ...    740474.0   
163     180558.0    182109.0    183630.0    185658.0  ...    552793.0   
187     118973.0    120408.0    121793.0    123521.0  ...    414301.0   
376     179344.0    181311.0    183272.0    185460.0  ...    563435.0   
506     156682.0    157906.0    159379.0    161537.0  ...    471998.0   
776     195425.0    197360.0    199502.0    201880.0  ...    613027.0   
1045    190305.0    192244.0    193904.0    196294.0  ...    623907.0   
1066    132951.0    134443.0    135886.0    137552.0  ...    434356.0   
1153    201344.0    203258.0    204758.0    206849.0  ...    598873.0   
1187    157614.0    159178.0    160669.0    162326.0  ...    492185.0   

      2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
102     743616.0    747228.0    758667.0    771280.0    783244.0    793192.0   
163     554208.0    556088.0    564176.0    572733.0    581232.0    588094.0   
187     417581.0    420311.0    427348.0    435113.0    442545.0    449057.0   
376     569282.0    574954.0    585692.0    597061.0    607935.0    617770.0   
506     474465.0    477300.0    484416.0    492241.0    499254.0    505231.0   
776     618232.0    623575.0    635420.0    647481.0    659269.0    668892.0   
1045    622980.0    623650.0    632260.0    642479.0    653340.0    663062.0   
1066    437932.0    441782.0    450028.0    458706.0    467239.0    474721.0   
1153    600673.0    602401.0    610556.0    619303.0    628203.0    635282.0   
1187    495516.0    498969.0    506790.0    515216.0    524209.0    532311.0   

      2022-05-31  2022-06-30  2022-07-31  
102     803220.0    810269.0    808449.0  
163     595136.0    599048.0    597289.0  
187     455644.0    460070.0    458351.0  
376     626599.0    632543.0    629973.0  
506     511299.0    515667.0    514395.0  
776     678421.0    684692.0    682833.0  
1045    671655.0    675421.0    670954.0  
1066    482443.0    487579.0    486112.0  
1153    642657.0    647330.0    647295.0  
1187    539111.0    543416.0    541992.0  

[10 rows x 272 columns]

 nulls? if so, how many?

In [17]:
sac.isna().sum().sum()

135

 nulls? which columns?

In [18]:
sac.isna().any()

RegionName    False
2000-01-31     True
2000-02-29     True
2000-03-31     True
2000-04-30     True
              ...  
2022-03-31    False
2022-04-30    False
2022-05-31    False
2022-06-30    False
2022-07-31    False
Length: 272, dtype: bool

 nulls? which rows?

In [19]:
sac[sac.isna().any(axis=1)]

RegionName  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
4720        95824     96280.0     96456.0     96817.0     97323.0     98136.0   
9318        95742         NaN         NaN         NaN         NaN         NaN   
11120       95832    127229.0    127814.0    128346.0    129255.0    129962.0   
14695       95655         NaN         NaN         NaN         NaN         NaN   

       2000-06-30  2000-07-31  2000-08-31  2000-09-30  ...  2021-10-31  \
4720      99043.0    100753.0    102337.0    103864.0  ...    350709.0   
9318          NaN         NaN         NaN         NaN  ...    615853.0   
11120    130884.0    131874.0    133181.0    134297.0  ...    390735.0   
14695         NaN         NaN         NaN         NaN  ...    555334.0   

       2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
4720     352093.0    353775.0    359113.0    365366.0    371287.0    376263.0   
9318     621054.0    626863.0    637645.0    649459.0    660568.0    671286.0   
11120    392838.0    394673.0    400416.0    406873.0    413608.0    419317.0   
14695    558488.0    562731.0    571954.0    582411.0    591254.0    598666.0   

       2022-05-31  2022-06-30  2022-07-31  
4720     381388.0    385475.0    385612.0  
9318     681191.0    688780.0    687901.0  
11120    424165.0    427472.0    427186.0  
14695    606136.0    611227.0    609348.0  

[4 rows x 272 columns]

 nulls. closer look at one col with nulls. zipCode 95655 (mather ca)

In [20]:
print(sac.loc[[14695]])

       RegionName  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
14695       95655         NaN         NaN         NaN         NaN         NaN   

       2000-06-30  2000-07-31  2000-08-31  2000-09-30  ...  2021-10-31  \
14695         NaN         NaN         NaN         NaN  ...    555334.0   

       2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
14695    558488.0    562731.0    571954.0    582411.0    591254.0    598666.0   

       2022-05-31  2022-06-30  2022-07-31  
14695    606136.0    611227.0    609348.0  

[1 rows x 272 columns]


 nulls. closer look at one col with nulls. zipCode 95742 (rancho coradova ca).
 which is right next to mather ca

In [21]:
print(sac.loc[[9318]])

      RegionName  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
9318       95742         NaN         NaN         NaN         NaN         NaN   

      2000-06-30  2000-07-31  2000-08-31  2000-09-30  ...  2021-10-31  \
9318         NaN         NaN         NaN         NaN  ...    615853.0   

      2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
9318    621054.0    626863.0    637645.0    649459.0    660568.0    671286.0   

      2022-05-31  2022-06-30  2022-07-31  
9318    681191.0    688780.0    687901.0  

[1 rows x 272 columns]


 mather and rancho cordova zip codes are focus zip codes, but most of the nans are in those columns up until 2010. i'd have t0 drop 1/2 the columns. instead, I will examine those zips later
*  drop zip codes

In [22]:
sac=sac.drop(9318)

In [23]:
sac=sac.drop(14695)

 nulls. what do we have left?

In [24]:
sac.isna().sum().sum()

2

 let's look at remaining nans

In [25]:
sac2 = sac[sac.isna().any(axis=1)]
sac2

RegionName  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
4720        95824     96280.0     96456.0     96817.0     97323.0     98136.0   
11120       95832    127229.0    127814.0    128346.0    129255.0    129962.0   

       2000-06-30  2000-07-31  2000-08-31  2000-09-30  ...  2021-10-31  \
4720      99043.0    100753.0    102337.0    103864.0  ...    350709.0   
11120    130884.0    131874.0    133181.0    134297.0  ...    390735.0   

       2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
4720     352093.0    353775.0    359113.0    365366.0    371287.0    376263.0   
11120    392838.0    394673.0    400416.0    406873.0    413608.0    419317.0   

       2022-05-31  2022-06-30  2022-07-31  
4720     381388.0    385475.0    385612.0  
11120    424165.0    427472.0    427186.0  

[2 rows x 272 columns]

 since only 2 nans and they are not next to each other lets ffill

In [26]:
sac.ffill(axis=0, inplace=True, limit=2)

 get a count of nans

In [27]:
sac.isna().sum().sum()

0

 quick last check of df

In [28]:
sac.head()

RegionName  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
102       95630    235832.0    236836.0    237997.0    240134.0    242395.0   
163       95608    174288.0    175027.0    175987.0    177583.0    179274.0   
187       95823    114478.0    114963.0    115594.0    116743.0    117922.0   
376       95758    172491.0    173242.0    174209.0    175923.0    177760.0   
506       95670    151622.0    152162.0    152938.0    154300.0    155740.0   

     2000-06-30  2000-07-31  2000-08-31  2000-09-30  ...  2021-10-31  \
102    244414.0    247327.0    250195.0    253888.0  ...    740474.0   
163    180558.0    182109.0    183630.0    185658.0  ...    552793.0   
187    118973.0    120408.0    121793.0    123521.0  ...    414301.0   
376    179344.0    181311.0    183272.0    185460.0  ...    563435.0   
506    156682.0    157906.0    159379.0    161537.0  ...    471998.0   

     2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  2022-04-30  \
102    743616.0    747228.0    758667.0    771280.0    783244.0    793192.0   
163    554208.0    556088.0    564176.0    572733.0    581232.0    588094.0   
187    417581.0    420311.0    427348.0    435113.0    442545.0    449057.0   
376    569282.0    574954.0    585692.0    597061.0    607935.0    617770.0   
506    474465.0    477300.0    484416.0    492241.0    499254.0    505231.0   

     2022-05-31  2022-06-30  2022-07-31  
102    803220.0    810269.0    808449.0  
163    595136.0    599048.0    597289.0  
187    455644.0    460070.0    458351.0  
376    626599.0    632543.0    629973.0  
506    511299.0    515667.0    514395.0  

[5 rows x 272 columns]

 calculate percent increase from yr to yr in sac
*  create df and define func (function)
*  create copy 
   *  copy() method creates a copy of an existing list

In [29]:
def calculate_percent_increase(x1, x2):
    return ((x2-x1)/x1)*100

In [30]:
sac_percent = sac.copy()
for i in range(sac_percent.shape[1]-1,1, -1):
    prior = sac_percent.iloc[:, i-1]
    current = sac_percent.iloc[:, i] 
    sac_percent.iloc[:, i] = calculate_percent_increase(prior, current)

 check df  percent increase

In [42]:
sac_percent.head()

95630          95608          95823          95758  \
01/2000  235832.000000  174288.000000  114478.000000  172491.000000   
02/2000       0.425727       0.424011       0.423662       0.435385   
03/2000       0.490213       0.548487       0.548872       0.558179   
04/2000       0.897910       0.906885       0.993996       0.983876   
05/2000       0.941558       0.952231       1.009911       1.044207   

                 95670          95624          95831          95828  \
01/2000  151622.000000  188868.000000  183734.000000  127990.000000   
02/2000       0.356149       0.382807       0.426160       0.443003   
03/2000       0.509983       0.459410       0.577183       0.462052   
04/2000       0.890557       0.849513       0.903105       0.934565   
05/2000       0.933247       0.931383       0.963911       1.011829   

                 95628          95610  ...          95811          95832  \
01/2000  194236.000000  151719.000000  ...  157097.000000  127229.000000   
02/2000       0.468502       0.382286  ...       0.253983       0.459801   
03/2000       0.506288       0.458966  ...       0.303500       0.416230   
04/2000       0.866245       0.908509  ...       0.281059       0.708242   
05/2000       0.997306       1.062259  ...       0.674797       0.546981   

                 95683          95693          95626          95690  \
01/2000  242949.000000  283204.000000  136766.000000  170603.000000   
02/2000       0.359335       0.441025       0.127956       0.041617   
03/2000       0.139036       0.313233       0.439605       1.637625   
04/2000       0.507862       0.917489       0.900082       1.373156   
05/2000       0.497553       1.070627       0.897817       1.986341   

                 95641          95638          95615          95830  
01/2000  178082.000000  219053.000000  172608.000000  312427.000000  
02/2000       0.251008       0.406751       0.193502       0.015684  
03/2000      -0.040330       0.186866       1.080709       0.469476  
04/2000      -0.874160       0.880851       1.089176       0.577812  
05/2000      -0.667620       1.365297       1.217780       0.438945  

[5 rows x 48 columns]

 transpose data
*  changing to wide format

In [32]:
sac_percent = sac_percent.set_index('RegionName').transpose()

rename sac df before transposing

In [33]:
sac_price = sac

In [34]:
sac_price = sac_price.set_index('RegionName').transpose()

In [35]:
sac_percent.head(3)

RegionName          95630          95608          95823          95758  \
2000-01-31  235832.000000  174288.000000  114478.000000  172491.000000   
2000-02-29       0.425727       0.424011       0.423662       0.435385   
2000-03-31       0.490213       0.548487       0.548872       0.558179   

RegionName          95670          95624          95831          95828  \
2000-01-31  151622.000000  188868.000000  183734.000000  127990.000000   
2000-02-29       0.356149       0.382807       0.426160       0.443003   
2000-03-31       0.509983       0.459410       0.577183       0.462052   

RegionName          95628          95610  ...          95811          95832  \
2000-01-31  194236.000000  151719.000000  ...  157097.000000  127229.000000   
2000-02-29       0.468502       0.382286  ...       0.253983       0.459801   
2000-03-31       0.506288       0.458966  ...       0.303500       0.416230   

RegionName          95683          95693          95626          95690  \
2000-01-31  242949.000000  283204.000000  136766.000000  170603.000000   
2000-02-29       0.359335       0.441025       0.127956       0.041617   
2000-03-31       0.139036       0.313233       0.439605       1.637625   

RegionName          95641          95638          95615          95830  
2000-01-31  178082.000000  219053.000000  172608.000000  312427.000000  
2000-02-29       0.251008       0.406751       0.193502       0.015684  
2000-03-31      -0.040330       0.186866       1.080709       0.469476  

[3 rows x 48 columns]

In [36]:
sac_price.head(3)

RegionName     95630     95608     95823     95758     95670     95624  \
2000-01-31  235832.0  174288.0  114478.0  172491.0  151622.0  188868.0   
2000-02-29  236836.0  175027.0  114963.0  173242.0  152162.0  189591.0   
2000-03-31  237997.0  175987.0  115594.0  174209.0  152938.0  190462.0   

RegionName     95831     95828     95628     95610  ...     95811     95832  \
2000-01-31  183734.0  127990.0  194236.0  151719.0  ...  157097.0  127229.0   
2000-02-29  184517.0  128557.0  195146.0  152299.0  ...  157496.0  127814.0   
2000-03-31  185582.0  129151.0  196134.0  152998.0  ...  157974.0  128346.0   

RegionName     95683     95693     95626     95690     95641     95638  \
2000-01-31  242949.0  283204.0  136766.0  170603.0  178082.0  219053.0   
2000-02-29  243822.0  284453.0  136941.0  170674.0  178529.0  219944.0   
2000-03-31  244161.0  285344.0  137543.0  173469.0  178457.0  220355.0   

RegionName     95615     95830  
2000-01-31  172608.0  312427.0  
2000-02-29  172942.0  312476.0  
2000-03-31  174811.0  313943.0  

[3 rows x 48 columns]

 Let's convert the index to datetime!

*  strftime() function is used to convert to Index using specified date_format.
*   columns.name=none- names the index, by default it is None or blank, the columns are of type IndexArray

In [37]:
sac_percent.index = p.DatetimeIndex(sac_percent.index)
sac_percent.index = sac_percent.index.strftime('%m/%Y')
sac_percent.columns.name = None

In [38]:
sac_price.index = p.DatetimeIndex(sac_price.index)
sac_price.index = sac_price.index.strftime('%m/%Y')
sac_price.columns.name = None

 exporting 🥒


In [39]:
with open('pickle/timeseries_sac_percent.pkl', 'wb') as fp:
    pickle.dump(sac_percent, fp)

In [40]:
with open('pickle/timeseries_sac_price.pkl', 'wb') as fp:
    pickle.dump(sac_price, fp)

In [41]:
sac_percent.to_csv('timeseries_sac_percent.csv', header=True)